tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs

```diff
@@ -647,6 +712,71 @@ private async Task RunOnBehalfOfTestAsync(LabResponse labResponse)
             await confidentialApp.GetAccountsAsync().ConfigureAwait(false);
 #pragma warning restore CS0618 // Type or member is obsolete
             Assert.IsNull(userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+
+            //Run OBO again. Should get token from cache
+            authResult = await confidentialApp.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.IsTrue(!userCacheRecorder.LastAfterAccessNotificationArgs.IsApplicationCache);
+            Assert.IsTrue(userCacheRecorder.LastAfterAccessNotificationArgs.HasTokens);
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+            Assert.AreEqual(TokenSource.Cache, authResult.AuthenticationResultMetadata.TokenSource);
+
+            //Expire access tokens
+            TokenCacheHelper.ExpireAccessTokens(confidentialApp.UserTokenCacheInternal);
+
+            //Run OBO again. Should do token refresh since the AT is expired
+            authResult = await confidentialApp.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.IsTrue(!userCacheRecorder.LastAfterAccessNotificationArgs.IsApplicationCache);
+            Assert.IsTrue(userCacheRecorder.LastAfterAccessNotificationArgs.HasTokens);
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+            Assert.AreEqual(TokenSource.IdentityProvider, authResult.AuthenticationResultMetadata.TokenSource);
+
+            //creating second app with no refresh tokens
+            var atItems = await confidentialApp.UserTokenCacheInternal.GetAllAccessTokensAsync(true).ConfigureAwait(false);
+            var confidentialApp2 = ConfidentialClientApplicationBuilder
+                .Create(confidentialClientID)
+                .WithAuthority(new Uri(oboHost + authResult.TenantId), true)
+                .WithClientSecret(secret)
+                .WithTestLogging()
+                .BuildConcrete();
+
+            TokenCacheHelper.ExpireAndSaveAccessToken(confidentialApp2.UserTokenCacheInternal, atItems.FirstOrDefault());
+
+            //Should perform OBO flow since the access token is expired and the refresh token does not exist
```

Comment:
The bit I don't understand here is that if the access token is expired, then it's certainly because the token that was passed in the web API (that is therefore in the **userAssertion**) has itself expired. In other words, I don't know when this could happen in real life? Or is it testing "no cache" ?

What we'd want to test (and I don't know how difficult this is) is:
- We call OBO with a particular user assertion.
- we wait for 1h until the user assertion / OBO token expires
- we then call OBO again with this expired user assertion, which is, in this case only used as a key to find the refresh token, and we get a new OBO token. Note that given OBO is used in a web API, the user assertion would come necessarily from the web api itself part of a long lived process, as if it was coming from a client of the Web API (calling the web API with an expired token), the incoming token would never have been validated by the web API (as it was expired)

Reply:
And what if the refresh token is no longer valid, ex. something on the resource side has changed, how do we revoke the RT to force a new AT/RT pair?

Reply:
@jmprieur the scenario you describe is tested here
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/161411bf059185d8b4e20d7cae0d6c3d29b2da13/tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs#L728-L734

The AT is expired by updating the exp on the MsalAccessTokenCacheItem and then the RT is used to refresh the OBO tokens.

This specific section here is testing this scenario which is a requirement from the task
"E2E Test: OBO with expired AT and missing RT in cache [S]"

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/161411bf059185d8b4e20d7cae0d6c3d29b2da13/tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs#L743-L757

If you are saying that it is not possible to reach a state where the AT is expired and no RT is available then perhaps the requirement is not needed.

@bgavrilMS @neha-bhargava

Reply:
@jmprieur - I don't think MSAL can't handle this scenario ? I think a middleware web api would need to hold on to the original assertion, even if it expires, and keep asking MSAL for a token. Are you suggesting an API like 

```AcquireTokenOnBehalfOf(s_scopes, userAssertionHashString)``` ?

Reply:
@jennyf19 - MSAL would throw an MsalUiRequiredException, which the middleware web API needs to handle by sending a 401 or by stopping. 

If the upstream web api receives the 401, it should send a new user assertion A2. The middleware web api calls `AcquireTokenOBO(scope, A2)` which in turn: 

- looks for an access token associated with A2 (none is found)
- looks for a refresh token associated with A2 (none is found)
- calls ESTS with grant_type = OBO + A2
- ESTS responds with AT + RT
- MSAL stores the AT and RT in the cache and associates them with A2

If you use a partitioned cache, the old AT / RT are simply left to rot. If you'd be (wrongly) using a flat cache, the old AT / RT would be overriden.